File: Beep.OilandGas.PPDM39.DataManagement.Tools/ConsolidateSeedData.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Beep.OilandGas.PPDM39.DataManagement.Tools
{
    public class ConsolidateSeedData
    {
        public static void Run(string[] args)
        {
            Console.WriteLine("========================================");
            Console.WriteLine("PPDM Seed Data Consolidation");
            Console.WriteLine("========================================");
            Console.WriteLine();

            // Get solution root (go up from Tools project to solution root)
            var solutionRoot = Path.GetFullPath(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "..", "..", "..", ".."));
            
            var csvDataPath = Path.Combine(solutionRoot, "Beep.OilandGas.PPDM39.DataManagement", "Core", "SeedData", "PPDMCSVData.json");
            var referenceDataPath = Path.Combine(solutionRoot, "Beep.OilandGas.PPDM39.DataManagement", "SeedData", "Templates", "PPDMReferenceData.json");
            
            // Allow override via command line arguments
            if (args.Length >= 1 && !string.IsNullOrWhiteSpace(args[0]))
            {
                csvDataPath = args[0];
            }
            if (args.Length >= 2 && !string.IsNullOrWhiteSpace(args[1]))
            {
                referenceDataPath = args[1];
            }

            Console.WriteLine("Configuration:");
            Console.WriteLine($"  CSV Data Path: {csvDataPath}");
            Console.WriteLine($"  Reference Data Path: {referenceDataPath}");
            Console.WriteLine();

            if (!File.Exists(csvDataPath))
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine($"ERROR: CSV data file not found: {csvDataPath}");
                Console.ResetColor();
                return;
            }
            
            if (!File.Exists(referenceDataPath))
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine($"ERROR: Reference data file not found: {referenceDataPath}");
                Console.ResetColor();
                return;
            }

            try
            {
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine("Loading CSV data...");
                Console.ResetColor();
                var csvData = JsonSerializer.Deserialize<JsonObject>(File.ReadAllText(csvDataPath));
                
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine("Loading reference data...");
                Console.ResetColor();
                var referenceData = JsonSerializer.Deserialize<JsonObject>(File.ReadAllText(referenceDataPath));
                
                // Process CSV data
                Console.WriteLine("Processing CSV data...");
                var csvTables = ProcessCsvData(csvData);
                Console.WriteLine($"Found {csvTables.Count} unique tables in CSV data");
                
                // Get existing tables
                var existingTables = GetExistingTableNames(referenceData);
                Console.WriteLine($"Found {existingTables.Count} existing tables in reference data");
                Console.WriteLine();
                
                // Consolidate
                var newTables = new List<string>();
                var updatedTables = new List<string>();
                
                Console.WriteLine("Consolidating data...");
                foreach (var kvp in csvTables)
                {
                    var tableName = kvp.Key;
                    var csvItems = kvp.Value;
                    
                    if (existingTables.Contains(tableName))
                    {
                        // Add new items to existing table
                        var added = AddToExistingTable(referenceData, tableName, csvItems);
                        if (added > 0)
                        {
                            updatedTables.Add(tableName);
                            Console.ForegroundColor = ConsoleColor.Yellow;
                            Console.WriteLine($"  Added {added} new items to {tableName}");
                            Console.ResetColor();
                        }
                    }
                    else
                    {
                        // Add new table
                        AddNewTable(referenceData, tableName, csvItems);
                        newTables.Add(tableName);
                        Console.ForegroundColor = ConsoleColor.Yellow;
                        Console.WriteLine($"  Added new table {tableName} with {csvItems.Count} items");
                        Console.ResetColor();
                    }
                }
                
                // Update description
                if (referenceData["description"] != null)
                {
                    referenceData["description"] = "PPDM standard reference tables (RA_* tables) with comprehensive standard values. This file contains consolidated seed data from PPDMCSVData.json and additional priority tables.";
                }
                
                Console.WriteLine();
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine("========================================");
                Console.WriteLine("Consolidation Summary");
                Console.WriteLine("========================================");
                Console.ResetColor();
                Console.WriteLine($"  New tables added: {newTables.Count}");
                Console.WriteLine($"  Existing tables updated: {updatedTables.Count}");
                
                // Save
                Console.WriteLine();
                Console.WriteLine("Saving consolidated data...");
                var options = new JsonSerializerOptions { WriteIndented = true };
                File.WriteAllText(referenceDataPath, JsonSerializer.Serialize(referenceData, options));
                
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"\nSuccessfully saved to: {referenceDataPath}");
                Console.ResetColor();
            }
            catch (Exception ex)
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine("========================================");
                Console.WriteLine("ERROR: Consolidation failed!");
                Console.WriteLine("========================================");
                Console.ResetColor();
                Console.WriteLine($"Error: {ex.Message}");
                Console.WriteLine();
                Console.WriteLine("Stack trace:");
                Console.WriteLine(ex.StackTrace);
                throw;
            }
        }
        
        private static Dictionary<string, List<JsonObject>> ProcessCsvData(JsonObject csvData)
        {
            var tables = new Dictionary<string, List<JsonObject>>();
            
            foreach (var entry in csvData)
            {
                if (entry.Value is not JsonObject entryObj) continue;
                
                var tableName = entryObj["tableName"]?.GetValue<string>();
                if (string.IsNullOrEmpty(tableName)) continue;
                
                // Map R_* to RA_*
                var mappedTableName = tableName.StartsWith("R_") ? "RA_" + tableName.Substring(2) : tableName;
                
                var headers = entryObj["headers"]?.AsArray().Select(h => h?.GetValue<string>()).ToArray();
                var rows = entryObj["rows"]?.AsArray();
                var fileName = entryObj["fileName"]?.GetValue<string>() ?? entry.Key;
                
                if (headers == null || rows == null) continue;
                
                var items = new List<JsonObject>();
                
                foreach (var rowNode in rows)
                {
                    if (rowNode is not JsonArray rowArray) continue;
                    
                    var row = rowArray.Select(r => r?.GetValue<string>() ?? "").ToArray();
                    if (row.Length != headers.Length) continue;
                    
                    // Skip header rows
                    if (row.Length > 0 && row[0].StartsWith("(")) continue;
                    
                    var item = ConvertRowToJson(row, headers, mappedTableName, fileName);
                    if (item != null)
                    {
                        items.Add(item);
                    }
                }
                
                if (items.Count > 0)
                {
                    if (!tables.ContainsKey(mappedTableName))
                    {
                        tables[mappedTableName] = new List<JsonObject>();
                    }
                    tables[mappedTableName].AddRange(items);
                }
            }
            
            return tables;
        }
        
        private static JsonObject ConvertRowToJson(string[] row, string[] headers, string tableName, string fileName)
        {
            var rowDict = new Dictionary<string, string>();
            for (int i = 0; i < headers.Length && i < row.Length; i++)
            {
                rowDict[headers[i]] = row[i] ?? "";
            }
            
            var name1 = rowDict.GetValueOrDefault("NAME1", "").Trim();
            if (string.IsNullOrEmpty(name1)) return null;
            
            var result = new JsonObject();
            
            // Determine primary field
            var primaryField = DeterminePrimaryField(tableName);
            result[primaryField] = name1;
            
            // Extract STATUS_TYPE from filename
            var statusType = ExtractStatusType(fileName);
            if (statusType != null && tableName.Contains("STATUS"))
            {
                result["STATUS_TYPE"] = statusType;
            }
            
            // ALIAS fields
            var alias = rowDict.GetValueOrDefault("ALIAS", "").Trim();
            var name4 = rowDict.GetValueOrDefault("NAME4", "").Trim();
            
            if (!string.IsNullOrEmpty(alias))
            {
                result["ALIAS_ID"] = alias;
                result["ALIAS_SHORT_NAME"] = alias;
            }
            else if (!string.IsNullOrEmpty(name4))
            {
                result["ALIAS_ID"] = name4;
                result["ALIAS_SHORT_NAME"] = name4;
            }
            
            if (!string.IsNullOrEmpty(name4))
            {
                result["ALIAS_LONG_NAME"] = name4;
            }
            else if (!string.IsNullOrEmpty(name1))
            {
                result["ALIAS_LONG_NAME"] = name1;
            }
            
            if (!result.ContainsKey("ALIAS_SHORT_NAME"))
            {
                result["ALIAS_SHORT_NAME"] = name1;
            }
            
            // ABBREVIATION
            var name2 = rowDict.GetValueOrDefault("NAME2", "").Trim();
            var name3 = rowDict.GetValueOrDefault("NAME3", "").Trim();
            if (!string.IsNullOrEmpty(name2))
            {
                result["ABBREVIATION"] = name2;
            }
            else if (!string.IsNullOrEmpty(name3))
            {
                result["ABBREVIATION"] = name3;
            }
            else if (!string.IsNullOrEmpty(alias))
            {
                result["ABBREVIATION"] = alias.Length > 4 ? alias.Substring(0, 4).ToUpper() : alias.ToUpper();
            }
            else
            {
                result["ABBREVIATION"] = name1.Length > 4 ? name1.Substring(0, 4).ToUpper() : name1.ToUpper();
            }
            
            // ACTIVE_IND
            var valueStatus = rowDict.GetValueOrDefault("VALUE_STATUS", "").Trim();
            result["ACTIVE_IND"] = valueStatus.ToLower().Contains("adopted") ? "Y" : 
                                  valueStatus.ToLower().Contains("deprecated") ? "N" : "Y";
            
            result["PREFERRED_IND"] = "Y";
            result["ORIGINAL_IND"] = "Y";
            
            // SOURCE
            var source = rowDict.GetValueOrDefault("SOURCE", "").Trim();
            result["SOURCE"] = string.IsNullOrEmpty(source) ? "PPDM" : source;
            
            return result;
        }
        
        private static string DeterminePrimaryField(string tableName)
        {
            var lower = tableName.ToLower();
            if (lower.Contains("status") && !lower.Contains("type"))
                return "STATUS";
            if (lower.Contains("cost_type"))
                return "COST_TYPE";
            if (lower.Contains("completion_type"))
                return "COMPLETION_TYPE";
            if (lower.Contains("completion_status_type"))
                return "COMPLETION_STATUS_TYPE";
            if (lower.Contains("property_type"))
                return "PROPERTY_TYPE";
            if (lower.Contains("lease_type"))
                return "LEASE_TYPE";
            if (lower.Contains("production_method"))
                return "PRODUCTION_METHOD";
            if (lower.Contains("production_status"))
                return "STATUS";
            if (lower.Contains("production_type"))
                return "PRODUCTION_TYPE";
            if (lower.Contains("equipment_type"))
                return "EQUIPMENT_TYPE";
            if (lower.Contains("equipment_status"))
                return "STATUS";
            if (lower.Contains("facility_type"))
                return "FACILITY_TYPE";
            if (lower.Contains("drilling_method"))
                return "DRILLING_METHOD";
            if (lower.Contains("drilling_type"))
                return "DRILLING_TYPE";
            if (lower.Contains("completion_method"))
                return "COMPLETION_METHOD";
            if (lower.Contains("reservoir_type"))
                return "RESERVOIR_TYPE";
            if (lower.Contains("formation_type"))
                return "FORMATION_TYPE";
            if (lower.Contains("lithology_type"))
                return "LITHOLOGY_TYPE";
            if (lower.Contains("allocation_type"))
                return "ALLOCATION_TYPE";
            if (lower.Contains("account_proc_type"))
                return "ACCOUNT_PROC_TYPE";
            if (lower.Contains("allowable_expense"))
                return "ALLOWABLE_EXPENSE";
            if (lower.Contains("activity_type"))
                return "ACTIVITY_TYPE";
            if (lower.Contains("activity_set_type"))
                return "ACTIVITY_SET_TYPE";
            if (lower.Contains("anl_method_set_type"))
                return "ANL_METHOD_SET_TYPE";
            if (lower.Contains("anl_confidence_type"))
                return "ANL_CONFIDENCE_TYPE";
            if (lower.Contains("anl_repeatability"))
                return "ANL_REPEATABILITY";
            if (lower.Contains("well_status_type"))
                return "STATUS_TYPE";
            if (lower.Contains("type"))
                return "TYPE";
            if (lower.Contains("uom") || lower.Contains("unit_of_measure"))
                return "UOM";
            if (lower.Contains("row_quality"))
                return "ROW_QUALITY";
            if (lower.Contains("source") && !lower.Contains("account"))
                return "SOURCE";
            if (lower.Contains("accounting_method"))
                return "ACCOUNTING_METHOD";
            return "NAME";
        }
        
        private static string ExtractStatusType(string fileName)
        {
            if (fileName.Contains("_PPDM"))
            {
                var parts = fileName.Split(new[] { "_PPDM" }, StringSplitOptions.None)[0].Split('_');
                if (parts.Length > 1)
                {
                    return string.Join("_", parts);
                }
            }
            return null;
        }
        
        private static HashSet<string> GetExistingTableNames(JsonObject referenceData)
        {
            var tables = new HashSet<string>();
            if (referenceData["tables"] is JsonArray tablesArray)
            {
                foreach (var tableNode in tablesArray)
                {
                    if (tableNode is JsonObject tableObj && tableObj["tableName"] != null)
                    {
                        tables.Add(tableObj["tableName"].GetValue<string>());
                    }
                }
            }
            return tables;
        }
        
        private static int AddToExistingTable(JsonObject referenceData, string tableName, List<JsonObject> newItems)
        {
            if (referenceData["tables"] is not JsonArray tablesArray) return 0;
            
            foreach (var tableNode in tablesArray)
            {
                if (tableNode is not JsonObject tableObj) continue;
                if (tableObj["tableName"]?.GetValue<string>() != tableName) continue;
                
                var existingData = tableObj["data"] as JsonArray ?? new JsonArray();
                var existingKeys = new HashSet<string>();
                
                // Get existing keys
                foreach (var itemNode in existingData)
                {
                    if (itemNode is not JsonObject itemObj) continue;
                    var primaryField = DeterminePrimaryField(tableName);
                    var primaryValue = itemObj[primaryField]?.GetValue<string>() ?? "";
                    var statusType = itemObj["STATUS_TYPE"]?.GetValue<string>() ?? "";
                    var key = string.IsNullOrEmpty(statusType) ? primaryValue : $"{statusType}:{primaryValue}";
                    if (!string.IsNullOrEmpty(key))
                    {
                        existingKeys.Add(key);
                    }
                }
                
                // Add new items that aren't duplicates
                int added = 0;
                foreach (var newItem in newItems)
                {
                    var primaryField = DeterminePrimaryField(tableName);
                    var primaryValue = newItem[primaryField]?.GetValue<string>() ?? "";
                    var statusType = newItem["STATUS_TYPE"]?.GetValue<string>() ?? "";
                    var key = string.IsNullOrEmpty(statusType) ? primaryValue : $"{statusType}:{primaryValue}";
                    
                    if (!string.IsNullOrEmpty(key) && !existingKeys.Contains(key))
                    {
                        existingData.Add(newItem);
                        existingKeys.Add(key);
                        added++;
                    }
                }
                
                tableObj["data"] = existingData;
                return added;
            }
            
            return 0;
        }
        
        private static void AddNewTable(JsonObject referenceData, string tableName, List<JsonObject> items)
        {
            if (referenceData["tables"] is not JsonArray tablesArray)
            {
                tablesArray = new JsonArray();
                referenceData["tables"] = tablesArray;
            }
            
            var dataArray = new JsonArray();
            foreach (var item in items)
            {
                dataArray.Add(item);
            }
            
            var newTable = new JsonObject
            {
                ["tableName"] = tableName,
                ["description"] = $"Reference data for {tableName}",
                ["data"] = dataArray
            };
            
            tablesArray.Add(newTable);
        }
    }
}

File: Beep.OilandGas.PPDM39.DataManagement/consolidate_seed_data.py
#!/usr/bin/env python3
"""
Script to consolidate PPDM seed data from PPDMCSVData.json into PPDMReferenceData.json
"""
import json
import os
from collections import defaultdict
from typing import Dict, List, Any, Set

def load_json(file_path: str) -> Dict:
    """Load JSON file"""
    with open(file_path, 'r', encoding='utf-8') as f:
        return json.load(f)

def save_json(file_path: str, data: Dict, indent: int = 2):
    """Save JSON file with formatting"""
    with open(file_path, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=indent, ensure_ascii=False)

def map_table_name(table_name: str) -> str:
    """Map R_* table names to RA_* table names"""
    if table_name.startswith('R_'):
        return 'RA_' + table_name[2:]
    return table_name

def extract_status_type_from_filename(file_name: str) -> str:
    """Extract STATUS_TYPE from filename (e.g., Life_Cycle_PPDM -> Life_Cycle)"""
    if '_PPDM' in file_name:
        parts = file_name.split('_PPDM')[0].split('_')
        if len(parts) > 1:
            return '_'.join(parts)
    return None

def convert_value_status_to_active_ind(value_status: str) -> str:
    """Convert VALUE_STATUS to ACTIVE_IND"""
    if not value_status:
        return "Y"
    value_status_lower = value_status.lower()
    if "adopted" in value_status_lower:
        return "Y"
    elif "deprecated" in value_status_lower:
        return "N"
    return "Y"

def determine_primary_field_name(table_name: str) -> str:
    """Determine the primary field name based on table name"""
    table_lower = table_name.lower()
    if 'status' in table_lower:
        return 'STATUS'
    elif 'type' in table_lower:
        if 'well_status_type' in table_lower:
            return 'STATUS_TYPE'
        elif 'cost_type' in table_lower:
            return 'COST_TYPE'
        elif 'completion_type' in table_lower:
            return 'COMPLETION_TYPE'
        elif 'completion_status_type' in table_lower:
            return 'COMPLETION_STATUS_TYPE'
        elif 'property_type' in table_lower:
            return 'PROPERTY_TYPE'
        elif 'lease_type' in table_lower:
            return 'LEASE_TYPE'
        elif 'production_method' in table_lower:
            return 'PRODUCTION_METHOD'
        elif 'production_status' in table_lower:
            return 'STATUS'
        elif 'production_type' in table_lower:
            return 'PRODUCTION_TYPE'
        elif 'equipment_type' in table_lower:
            return 'EQUIPMENT_TYPE'
        elif 'equipment_status' in table_lower:
            return 'STATUS'
        elif 'facility_type' in table_lower:
            return 'FACILITY_TYPE'
        elif 'drilling_method' in table_lower:
            return 'DRILLING_METHOD'
        elif 'drilling_type' in table_lower:
            return 'DRILLING_TYPE'
        elif 'completion_method' in table_lower:
            return 'COMPLETION_METHOD'
        elif 'reservoir_type' in table_lower:
            return 'RESERVOIR_TYPE'
        elif 'formation_type' in table_lower:
            return 'FORMATION_TYPE'
        elif 'lithology_type' in table_lower:
            return 'LITHOLOGY_TYPE'
        elif 'allocation_type' in table_lower:
            return 'ALLOCATION_TYPE'
        elif 'account_proc_type' in table_lower:
            return 'ACCOUNT_PROC_TYPE'
        elif 'allowable_expense' in table_lower:
            return 'ALLOWABLE_EXPENSE'
        elif 'activity_type' in table_lower:
            return 'ACTIVITY_TYPE'
        elif 'activity_set_type' in table_lower:
            return 'ACTIVITY_SET_TYPE'
        elif 'anl_method_set_type' in table_lower:
            return 'ANL_METHOD_SET_TYPE'
        elif 'anl_confidence_type' in table_lower:
            return 'ANL_CONFIDENCE_TYPE'
        elif 'anl_repeatability' in table_lower:
            return 'ANL_REPEATABILITY'
        return 'TYPE'
    elif 'method' in table_lower:
        return 'METHOD'
    elif 'uom' in table_lower or 'unit_of_measure' in table_lower:
        return 'UOM'
    elif 'row_quality' in table_lower:
        return 'ROW_QUALITY'
    elif 'source' in table_lower and 'account' not in table_lower:
        return 'SOURCE'
    elif 'accounting_method' in table_lower:
        return 'ACCOUNTING_METHOD'
    return 'NAME'

def convert_csv_row_to_json(row: List[str], headers: List[str], table_name: str, file_name: str) -> Dict[str, Any]:
    """Convert CSV row (array of strings) to JSON object"""
    if len(row) != len(headers):
        return None
    
    # Skip header rows or invalid rows
    if len(row) == 0 or (len(row) > 0 and row[0].startswith('(')):
        return None
    
    # Create mapping
    row_dict = dict(zip(headers, row))
    
    # Determine primary field name
    primary_field = determine_primary_field_name(table_name)
    
    # Extract STATUS_TYPE from filename if applicable
    status_type = extract_status_type_from_filename(file_name)
    
    # Build JSON object
    result = {}
    
    # Primary field (STATUS, TYPE, etc.)
    name1 = row_dict.get('NAME1', '').strip()
    if name1:
        result[primary_field] = name1
    
    # STATUS_TYPE (if applicable)
    if status_type and 'status' in table_name.lower():
        result['STATUS_TYPE'] = status_type
    
    # ALIAS fields
    alias = row_dict.get('ALIAS', '').strip()
    name4 = row_dict.get('NAME4', '').strip()
    
    if alias:
        result['ALIAS_ID'] = alias
        result['ALIAS_SHORT_NAME'] = alias
    elif name4:
        result['ALIAS_ID'] = name4
        result['ALIAS_SHORT_NAME'] = name4
    
    if name4:
        result['ALIAS_LONG_NAME'] = name4
    elif name1:
        result['ALIAS_LONG_NAME'] = name1
    
    if not result.get('ALIAS_SHORT_NAME'):
        result['ALIAS_SHORT_NAME'] = name1 if name1 else alias
    
    # ABBREVIATION
    name2 = row_dict.get('NAME2', '').strip()
    name3 = row_dict.get('NAME3', '').strip()
    if name2:
        result['ABBREVIATION'] = name2
    elif name3:
        result['ABBREVIATION'] = name3
    elif alias:
        # Use first few chars of alias as abbreviation
        result['ABBREVIATION'] = alias[:4].upper() if len(alias) > 4 else alias.upper()
    else:
        result['ABBREVIATION'] = name1[:4].upper() if len(name1) > 4 else name1.upper()
    
    # ACTIVE_IND
    value_status = row_dict.get('VALUE_STATUS', '').strip()
    result['ACTIVE_IND'] = convert_value_status_to_active_ind(value_status)
    
    # PREFERRED_IND and ORIGINAL_IND
    result['PREFERRED_IND'] = "Y"
    result['ORIGINAL_IND'] = "Y"
    
    # SOURCE
    source = row_dict.get('SOURCE', '').strip()
    if source:
        result['SOURCE'] = source
    else:
        result['SOURCE'] = "PPDM"
    
    # Only return if we have at least a primary field value
    if result.get(primary_field):
        return result
    
    return None

def process_csv_data(csv_data_path: str) -> Dict[str, List[Dict]]:
    """Process PPDMCSVData.json and convert to structured format"""
    csv_data = load_json(csv_data_path)
    
    # Group by mapped table name
    tables_data = defaultdict(list)
    
    for file_key, entry in csv_data.items():
        if not isinstance(entry, dict):
            continue
        
        table_name = entry.get('tableName', '')
        if not table_name:
            continue
        
        # Map table name
        mapped_table_name = map_table_name(table_name)
        
        headers = entry.get('headers', [])
        rows = entry.get('rows', [])
        file_name = entry.get('fileName', file_key)
        
        # Convert rows
        for row in rows:
            converted = convert_csv_row_to_json(row, headers, mapped_table_name, file_name)
            if converted:
                tables_data[mapped_table_name].append(converted)
    
    return dict(tables_data)

def get_existing_table_names(reference_data: Dict) -> Set[str]:
    """Get set of existing table names in PPDMReferenceData.json"""
    tables = reference_data.get('tables', [])
    return {table.get('tableName', '') for table in tables if table.get('tableName')}

def get_existing_data_keys(reference_data: Dict, table_name: str) -> Set[str]:
    """Get set of unique keys for existing data in a table (for duplicate detection)"""
    tables = reference_data.get('tables', [])
    for table in tables:
        if table.get('tableName') == table_name:
            data = table.get('data', [])
            # Create unique keys based on primary field
            keys = set()
            for item in data:
                # Use primary field value as key
                primary_field = determine_primary_field_name(table_name)
                primary_value = item.get(primary_field, '')
                status_type = item.get('STATUS_TYPE', '')
                if status_type:
                    keys.add(f"{status_type}:{primary_value}")
                else:
                    keys.add(primary_value)
            return keys
    return set()

def is_duplicate(new_item: Dict, existing_keys: Set[str], table_name: str) -> bool:
    """Check if new item is duplicate of existing data"""
    primary_field = determine_primary_field_name(table_name)
    primary_value = new_item.get(primary_field, '')
    status_type = new_item.get('STATUS_TYPE', '')
    
    if status_type:
        key = f"{status_type}:{primary_value}"
    else:
        key = primary_value
    
    return key in existing_keys

def consolidate_data(csv_data_path: str, reference_data_path: str, output_path: str):
    """Main consolidation function"""
    print("Loading CSV data...")
    csv_tables = process_csv_data(csv_data_path)
    print(f"Found {len(csv_tables)} tables in CSV data")
    
    print("Loading reference data...")
    reference_data = load_json(reference_data_path)
    existing_tables = get_existing_table_names(reference_data)
    print(f"Found {len(existing_tables)} existing tables in reference data")
    
    # Process each CSV table
    new_tables = []
    updated_tables = []
    
    for table_name, csv_items in csv_tables.items():
        if table_name in existing_tables:
            # Table exists, check for new data
            existing_keys = get_existing_data_keys(reference_data, table_name)
            new_items = [item for item in csv_items if not is_duplicate(item, existing_keys, table_name)]
            
            if new_items:
                # Find existing table and add new items
                for table in reference_data['tables']:
                    if table.get('tableName') == table_name:
                        table['data'].extend(new_items)
                        updated_tables.append(table_name)
                        print(f"  Added {len(new_items)} new items to {table_name}")
                        break
        else:
            # New table, add it
            new_table = {
                "tableName": table_name,
                "description": f"Reference data for {table_name}",
                "data": csv_items
            }
            reference_data['tables'].append(new_table)
            new_tables.append(table_name)
            print(f"  Added new table {table_name} with {len(csv_items)} items")
    
    # Update description
    reference_data['description'] = "PPDM standard reference tables (RA_* tables) with comprehensive standard values. This file contains consolidated seed data from PPDMCSVData.json and additional priority tables."
    
    print(f"\nSummary:")
    print(f"  New tables added: {len(new_tables)}")
    print(f"  Existing tables updated: {len(updated_tables)}")
    
    print(f"\nSaving to {output_path}...")
    save_json(output_path, reference_data)
    print("Done!")

if __name__ == '__main__':
    base_dir = os.path.dirname(os.path.abspath(__file__))
    csv_data_path = os.path.join(base_dir, 'Core', 'SeedData', 'PPDMCSVData.json')
    reference_data_path = os.path.join(base_dir, 'SeedData', 'Templates', 'PPDMReferenceData.json')
    output_path = reference_data_path  # Overwrite the original
    
    consolidate_data(csv_data_path, reference_data_path, output_path)

